### PR TITLE
opt: improve FDs for Intersect and Except variants

### PR DESCRIPTION
**Pull request by RaduBerinde**
We can pass through FDs from the left side on Except/ExceptAll; and we
can pass through FDs from both sides on Intersect/IntersectAll.

Release note: None

### CODE REVIEW

File: pkg/sql/opt/props/func_dep.go [hunk 1]
```diff
@@ -551,6 +551,23 @@ func (f *FuncDepSet) CopyFrom(fdset *FuncDepSet) {
 	f.hasKey = fdset.hasKey
 }
 
+// RemapFrom copies the given FD into this FD, remapping column IDs according to
+// the given function.
+func (f *FuncDepSet) RemapFrom(fdset *FuncDepSet, remap func(opt.ColumnID) opt.ColumnID) {
```
Comment (DrewKimball) [inline pkg/sql/opt/props/func_dep.go:556]: There's a function `TranslateColSet` in opt/colset.go that already has this sort of logic - you could just pass in the left and output ColLists here and call `TranslateColSet`

**Review by DrewKimball — APPROVED** [review]

**Review by rytaft — COMMENTED** [review]
Nice!

> We can pass through FDs from the left side on Intersect/IntersectAll,
> and we can pass through FDs from both sides on Except/ExceptAll.

Isn't it the opposite? (i.e., can pass from both sides for intersect but not except?) I think the code is already doing the right thing, just seems like the commit/PR message needs to be updated.

Might be worth adding an explicit test that this is the case, though.

> Reviewed 15 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-:-McBPUTaDDsicmih3_ut:b-yc2yjo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/props/func_dep.go, line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-McAtJ31QCIPWZuadR6o-r1-556:-McBUgfq8BPOKGqWsrPI:b-ul8utz) ([raw file](https://github.com/cockroachdb/cockroach/blob/491690a4566b54108245fe457150548d8b5646fa/pkg/sql/opt/props/func_dep.go#L556)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

There's a function `TranslateColSet` in opt/colset.go that already has this sort of logic - you could just pass in the left and output ColLists here and call `TranslateColSet`
</blockquote></details>

Agreed. I did something similar here with OrderingChoice: https://github.com/cockroachdb/cockroach/blob/970a932bebef1ecb5b900ed011705d6044148106/pkg/sql/opt/props/ordering_choice.go#L905

**Review by RaduBerinde — COMMENTED** [review]
Commit message fixed. There are some existing tests in logprops/set.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-:-McBZssD1OWbn6iNAodx:b6tqjql)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @rytaft)

___
*[pkg/sql/opt/props/func_dep.go, line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-McAtJ31QCIPWZuadR6o-r1-556:-McBaTPk6YGMKjdbBx5K:b6rkyve) ([raw file](https://github.com/cockroachdb/cockroach/blob/491690a4566b54108245fe457150548d8b5646fa/pkg/sql/opt/props/func_dep.go#L556)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Agreed. I did something similar here with OrderingChoice: https://github.com/cockroachdb/cockroach/blob/970a932bebef1ecb5b900ed011705d6044148106/pkg/sql/opt/props/ordering_choice.go#L905
</blockquote></details>

Done. I found that most current usecases of this function legitimately pass in columns that are not in the `from` list. I updated the comments and added a `Strict` version and converted some of the existing calls.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 8 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-:-McF-ClLBgka4YV5ePGO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

___
*[pkg/sql/opt/props/func_dep.go, line 556 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-McAtJ31QCIPWZuadR6o-r1-556:-McEyqc220N7tTjszeji:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/491690a4566b54108245fe457150548d8b5646fa/pkg/sql/opt/props/func_dep.go#L556)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Done. I found that most current usecases of this function legitimately pass in columns that are not in the `from` list. I updated the comments and added a `Strict` version and converted some of the existing calls.
</blockquote></details>

Thanks!

**Review by RaduBerinde — COMMENTED** [review]
TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66444#-:-McFCtX_BCO_ZpcAiRZk:b-tlx7ko)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)

**Pull request merged**